### PR TITLE
[SMTChecker] Unroll loops once

### PR DESCRIPTION
**Pull request by leonardoalt**
Currently the SMT checker unrolls the loop once and resets knowledge.
This PR:
- Keeps that as an execution branch (one unroll) which merges with the branch that skips the loop and does not clear previous knowledge.
- Moves the loop comment regarding knowledge erasing to a `SecondarySourceLocation` (does not show in tests).
- Adds more tests.

### CODE REVIEW

File: test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol [hunk 1]
```diff
@@ -5,4 +5,3 @@ contract C
 	function f(bool x) public pure { require(x); for (;x;) {} }
 }
 // ----
-// Warning: (98-99): For loop condition is always true.
```
Comment (axic) [inline test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol]: Why is this removed?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol]: Because this PR disables the triviality check for loop conditions, since now we don't just run once and erase all knowledge but rather consider the two branches (in loop and skip loop). Also, it is common that the first time a condition is evaluated during execution it is true, which might not be true at some point.

Comment (chriseth) [reply test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol]: But we should still warn about it, shouldn't we?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol]: But how can we be sure?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol]: In this case we know `x` is `true` for the first iteration, but we don't know it's always true without analyzing the body in an unbounded way.

Comment (chriseth) [reply test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol]: OK, I see what you mean. So the error for now had a high false-positive rate? Didn't we at least do something like clear everything that is touched in the loop body?

We should at least be able to infer whether it is always false.

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/functions/functions_trivial_condition_for.sol]: Yes, this PR keeps that clearing part, but it also considers two separate branches and merges them.
True, I'll keep the always false check then.

**Comment by leonardoalt** [comment]
@chriseth Updated. I had missed something in the constant check discussion, and now I see that the trivial condition case is indeed safe if we reset the touched variables. Keeping both checks.

File: test/libsolidity/smtCheckerTests/loops/for_1_fail.sol [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	function f(uint x) public pure {
+		require(x < 100);
+		for(uint i = 0; i < 10; ++i) {
+			x = x + 1;
+		}
+		assert(x < 14);
+	}
+}
+// ----
+// Warning: (140-145): Overflow (resulting value larger than 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) happens here
```
Comment (chriseth) [inline test/libsolidity/smtCheckerTests/loops/for_1_fail.sol]: This is a false positive, right? If so, please add a comment.

File: test/libsolidity/smtCheckerTests/loops/for_1.sol [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+	function f(uint x) public pure {
+		require(x < 100);
+		for(uint i = 0; i < 10; ++i) {
+			x = x + 1;
+		}
+		assert(x > 0);
+	}
+}
+// ----
+// Warning: (140-145): Overflow (resulting value larger than 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) happens here
```
Comment (chriseth) [inline test/libsolidity/smtCheckerTests/loops/for_1.sol]: Also false positive?

File: test/libsolidity/smtCheckerTests/loops/do_while_1_fail.sol [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+	function f(uint x) public pure {
+		require(x < 100);
+		do {
+			x = x + 1;
+		} while (x < 10);
+		assert(x < 14);
+	}
+}
+// ----
+// Warning: (114-119): Overflow (resulting value larger than 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) happens here
```
Comment (chriseth) [inline test/libsolidity/smtCheckerTests/loops/do_while_1_fail.sol]: False positive?

File: test/libsolidity/smtCheckerTests/loops/do_while_1.sol [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+		do {
+			x = x + 1;
+		} while (x < 1000);
+		assert(x > 0);
+	}
+}
+// ----
+// Warning: (114-119): Overflow (resulting value larger than 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) happens here
+// Warning: (110-119): Overflow (resulting value larger than 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) happens here
+// Warning: (145-158): Assertion violation happens here
```
Comment (chriseth) [inline test/libsolidity/smtCheckerTests/loops/do_while_1.sol]: False positive?

File: libsolidity/formal/SMTChecker.cpp [hunk 6]
```diff
@@ -838,19 +858,19 @@ void SMTChecker::checkCondition(
...
-			m_errorReporter.warning(_location, message.str() + loopComment);
+			m_errorReporter.warning(_location, message.str(), SecondarySourceLocation().append(loopComment, SourceLocation()));
 		}
 		break;
 	}
 	case smt::CheckResult::UNSATISFIABLE:
 		break;
 	case smt::CheckResult::UNKNOWN:
-		m_errorReporter.warning(_location, _description + " might happen here." + loopComment);
+		m_errorReporter.warning(_location, _description + " might happen here.", SecondarySourceLocation().append(loopComment, SourceLocation()));
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp:883]: How will that be displayed in the commandline compiler (it does not have a location)?

File: libsolidity/formal/SMTChecker.cpp [hunk 7]
```diff
@@ -193,13 +206,19 @@ bool SMTChecker::visit(ForStatement const& _node)
 	_node.body().accept(*this);
 	if (_node.loopExpression())
 		_node.loopExpression()->accept(*this);
-
 	m_interface->pop();
 
-	m_loopExecutionHappened = true;
+	auto indicesAfterLoop = copyVariableIndices();
+	// We reset the execution to before the loop
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp:222]: Can you add a comment about unrolling? You explain what is done but not why.

Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp:222]: The unrolling part is the same as before: we run it once. The difference is that before all the variables touched inside the loop body were reset after the loop, whereas now it merges with the branch that doesn't go into the loop, keeping the branch's knowledge

Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp:222]: Will add a comment

File: libsolidity/formal/SMTChecker.cpp (hunk 6 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp:883]: Just as a message in the end of the error, example:
```
../../test/libsolidity/smtCheckerTests/loops/while_loop_simple_5.sol:8:9: Warning: Assertion violation happens here
        assert(x == 7);
        ^------------^
  for:
  x = 0
  y = 0


Note that some information is erased after the execution of loops.
You can re-introduce information using require().
```

**Review by chriseth — APPROVED** [review]
